### ISSUE TITLE
keep port,addr as state variables, and allow `reconnect` otherwise client code will have to add this functionality

### ISSUE THREAD

**Issue by timotheecour** [rpclib/rpclib#117]

**Comment by sztomi** [comment]
This will be covered by a different feature that provides a way to register a callback for connection state changes and I'll probably add a reconnect call. In fact, I want to provide a default callback, because I think most people would want to reconnect anyway.

**Comment by sztomi** [comment]
@timotheecour, @dkozel:

This is my current idea for the connection state notification and reconnecting:

1. Callback type for connection state changes:

```cpp
using conn_state_cb = void (client&, client::connection_state);
```

2. client is extended with the following:

```cpp
template<typename F>
void set_connection_state_handler(F handler); // F is callable as conn_state_cb

static void default_conn_state_handler(client& c, connection_state state);

void reconnect(); // attempts connection if not connected, disconnects and connects if connected
```

By default, the `default_conn_state_handler` is registered and implements a graceful reconnection logic if the connections breaks.

Any thoughts?

**Comment by sztomi** [comment]
Also, @lovettchris I'd be interested in your opinion about the above idea.^

**Comment by lovettchris** [comment]
Sounds perfect.

**Comment by lovettchris** [comment]
Hey, can you also notify when connected? And if I call connect before server is up I don't want a long timeout before attempting reconnect, so what if connect and reconnect also take a connection timeout argument so I can control how quickly it makes the connection once the server is available...

**Comment by dkozel** [comment]
Hi @sztomi a few things come to mind.

What will the `default_conn_state_handler` do if the server is completely down?

Why does `reconnect` disconnect and reconnect if the connection is still valid? How can does it check if the connection is up? What happens if this call is not able to reconnect? Will the global timeout apply to this call? Having a return status would be useful, maybe connection_state or some other state indicator.

**Comment by sztomi** [comment]
@lovettchris yes, the callback will be called for all state changes, so you can react to any of those events.

>  And if I call connect before server is up I don't want a long timeout before attempting reconnect

There is no explicit `connect` function. Connection is initiated asynchronously when the client object is created (so no blocking while creating the client). `call` checks if there is already a connection and blocks until it is established. 

> a connection timeout argument so I can control how quickly it makes the connection once the server is available...

You will also be able to control reconnection frequency in your callback functor any way you wish. I'll probably add a new state for indicating that a connection attempt is in progress so it's easy to react to the right state changes.

**Comment by sztomi** [comment]
@dkozel 

> What will the default_conn_state_handler do if the server is completely down?

On the client side, we only get either a "connection reset" or "eof" when the TCP connection breaks. So we really can't take into account any other kind of server state in the default handler. That said, if your application does have a way to know more about the server state, then you can implement a handler that uses that information to act in a smarter way. The default handler will implement a reconnection logic that will be useful _for the majority_ of use-cases (i.e. reconnecting with increasing time intervals). Any other needs should be handled with a custom callback. Think of the default handler as a "batteries included" item. I'm adding it because I got feedback that people would expect rpclib to attempt reconnection on its own and I agree. But I don't want to force anyone into a specific reconnection logic because each application has unique constraints, so that's why there will be a way to set you own callback.

> Why does reconnect disconnect and reconnect if the connection is still valid? 

This is a language problem. People looking at the functions of client will see "reconnect" and may interpret as "re-create the existing connection". I think that allowing this behavior to be valid is better than throwing an exception, because it still won't introduce surprising behavior. But I'm open to suggestions.

> How can does it check if the connection is up?  What happens if this call is not able to reconnect?

My idea is that you shouldn't really check the connection state synchronously in the handler. Rather, you simply react to the states. E.g. when you see that the state changes to disconnected, you start a reconnection. As mentioned above, there will be a new state for "connection in progress", so a custom handler can react to failing the reconnection (e.g. you get a "connection in progress" -> disconnected state change).

> Will the global timeout apply to this call?

Good question. Maybe it would make sense to have `async_reconnect` and `reconnect` in the same vein as `call` and `async_call`. After the sync version returns, you can check the connection state of the client directly.

**Comment by sztomi** [comment]
Although the timeout does implicitly apply to the reconnection, since it applies to all calls and all calls waiting for connection are affected by the timeout.

**Comment by dkozel** [comment]
@sztomi 

> reconnecting with increasing time intervals

Thanks, this is very reasonable.

For `reconnect` I think that calling it when the connection state is not disconnected should be a no-op and simply return. Or have some test packet that it can send to the server to identify if the connection is actually alive or not.

My concern with the timeout and `reconnect` behavior is to avoid a situation where an application calls reconnect and it never returns because the server has gone away for some reason. Having an async_reconnect helps most of this because an application can then implement it's own timeout and error handling logic. There should be a way to end the reconnection attempt if the application decides that it no longer wants to try reconnecting to that server.

Thanks!

**Comment by lovettchris** [comment]
This is what I'm trying to avoid with control over a timeout:
a) launch client, it tries to auto-connect, but server is not ready, so it goes into some default TCP connection timeout - let's say that is 60 seconds.
b) server is ready 1 second later, but because the above is "waiting" for response from server that is not there it times out.  
c) So now I have to wait 59 seconds before I can call reconnect and get the actual connection I should have been able to get in 1 second.

Today when I run client and launch server 1 second later, it does not connect.  This is why I'm concerned.  If you can fix that so I don't have to worry about it then all the better.

**Comment by sztomi** [comment]
> For reconnect I think that calling it when the connection state is not disconnected should be a no-op and simply return. 

That was my first thought, but isn't that surprising behavior for people who expect it to recreate the existing connection?

> Or have some test packet

Can't do that unfortunately because it's not compatible with msgpack-rpc (you can do it on the application level though)

> My concern with the timeout and reconnect behavior is to avoid a situation where an application calls reconnect and it never returns because the server has gone away for some reason. 

Reconnect behavior is customizable in the callback. Say, if you want your client to retry ten times, but no more, then you can have a stateful functor as the callback and check the elapsed time. 

On that note, I'm liking the reconnect-async_reconnect pair more, because that allows applying the timeout right on the reconnect function. Let's see an example of the current state of the proposal:

## async option

```cpp
rpc::client client("localhost", 8080);
client.set_timeout(5000);

struct reconnect_handler_t {
    reconnect_handler_t() : retry_count_(0) {}
    void operator()(client& c, connection_state state) {
        if (rpc::is_disconnected(c) && retry_count_++ < MAX_TRIES) {
            c.async_reconnect();
        }
    }
    unsigned retry_count_;
};

client.set_connection_state_handler(reconnect_handler_t());

// rpc::is_disconnected(c) is a shorthand for `c.get_connection_state() == rpc::client::connection_state::disconnected`
```

## sync option

```cpp
rpc::client client("localhost", 8080);
client.set_timeout(5000);

client.set_connection_state_handler([](client& c, connection_state state) {
    while (retry_count_++ < MAX_TRIES) {
        try {
            if (rpc::is_disconnected(c)) {
                c.reconnect();
            }
        }
        catch(rpc::timeout&) {
            // handle timeout
        }
    }
});
```

I'm not sure about the latter one yet, since it uses exceptions for flow control, which is not great. It would be nice to allow users to avoid that.

**Comment by sztomi** [comment]
@lovettchris yes, this situation will be definitely fixed by this change (with the default handler, out-of-box, but you can customize it). 

To be clear, the last two examples posted are of the advanced usage (where you want to customize the behavior). Most users won't have to do anything, the default handler will strive for a connection no matter what.

**Comment by sztomi** [comment]
This is now in progress on the following branch: https://github.com/rpclib/rpclib/tree/policy

Any comments are welcome.
